packages/block-library/src/embed/edit.js

```diff
@@ -0,0 +1,382 @@
+import {
+	Button,
+	Placeholder,
+	Spinner,
+	SandBox,
+	IconButton,
+	Toolbar,
+	PanelBody,
+	ToggleControl,
+} from '@wordpress/components';
+import { createBlock } from '@wordpress/blocks';
+import { RichText, BlockControls, BlockIcon, InspectorControls } from '@wordpress/editor';
+
+export function getEmbedEditComponent( title, icon ) {
+	return class extends Component {
+		constructor() {
+			super( ...arguments );
+			this.switchBackToURLInput = this.switchBackToURLInput.bind( this );
+			this.setUrl = this.setUrl.bind( this );
+			this.maybeSwitchBlock = this.maybeSwitchBlock.bind( this );
+			this.getAttributesFromPreview = this.getAttributesFromPreview.bind( this );
+			this.setAttributesFromPreview = this.setAttributesFromPreview.bind( this );
+			this.setAspectRatioClassNames = this.setAspectRatioClassNames.bind( this );
+			this.getResponsiveHelp = this.getResponsiveHelp.bind( this );
+			this.toggleResponsive = this.toggleResponsive.bind( this );
+			this.handleIncomingPreview = this.handleIncomingPreview.bind( this );
+
+			this.state = {
+				editingURL: false,
+				url: this.props.attributes.url,
+			};
+
+			if ( this.props.preview ) {
+				this.handleIncomingPreview();
+			}
+		}
+
+		handleIncomingPreview() {
+			this.setAttributesFromPreview();
+			this.maybeSwitchBlock();
+		}
+
+		componentDidUpdate( prevProps ) {
+			const hasPreview = undefined !== this.props.preview;
+			const hadPreview = undefined !== prevProps.preview;
+			const switchedPreview = this.props.preview && this.props.attributes.url !== prevProps.attributes.url;
+			const switchedURL = this.props.attributes.url !== prevProps.attributes.url;
+
+			if ( ( switchedURL || ( hasPreview && ! hadPreview ) ) && this.maybeSwitchBlock() ) {
+				// Dont do anything if we are going to switch to a different block,
+				// and we've just changed the URL, or we've just received a preview.
+				return;
+			}
+
+			if ( ( hasPreview && ! hadPreview ) || switchedPreview ) {
+				if ( this.props.cannotEmbed ) {
+					// Can't embed this URL, and we've just received or switched the preview.
+					this.setState( { editingURL: true } );
+					return;
+				}
+				this.handleIncomingPreview();
+			}
+		}
+
+		getPhotoHtml( photo ) {
+			// 100% width for the preview so it fits nicely into the document, some "thumbnails" are
+			// acually the full size photo.
+			const photoPreview = <p><img src={ photo.thumbnail_url } alt={ photo.title } width="100%" /></p>;
+			return renderToString( photoPreview );
+		}
+
+		setUrl( event ) {
+			if ( event ) {
+				event.preventDefault();
+			}
+			const { url } = this.state;
+			const { setAttributes } = this.props;
+			this.setState( { editingURL: false } );
+			setAttributes( { url } );
+		}
+
+		/***
+		 * Switches to a different embed block type, based on the URL
+		 * and the HTML in the preview, if the preview or URL match a different block.
+		 *
+		 * @return {boolean} Whether the block was switched.
+		 */
+		maybeSwitchBlock() {
```

Comment:
I think this function is a bit problematic, as it has led to writing this elsewhere in the file:
```
if ( ( switchedURL || ( hasPreview && ! hadPreview ) ) && this.maybeSwitchBlock() ) {
```
and it's not great to have side-effects in an if statement. Making this purer in the functional sense would great. I think that's fairly easy to do, it could just return the result of `createBlock` instead of a boolean (or undefined if no block should be created), then the callsite can be responsible for calling onReplace on the result if there is one.

Naming it is the harder thing. The best I could come up with is `createSpecificEmbedBlock` or `createUpgradedEmbedBlock` (I'm bad at naming!).

Using it would be like:
```
const upgradedBlock = createUpgradedEmbedBlock( url, preview );
if ( upgradedBlock ) {
    this.props.onReplace( upgradedBlock );
}
```

createUpgradedEmbedBlock could even be moved to utils if there ends up being no use of `this` in its implementation, and then it can be unit tested 😄

Reply:
Heh yeah there are a few peeps who don't like that `maybe` prefix too :)

I like the sound of `createUpgradedEmbedBlock` that returns `undefined` if there was nothing to upgrade to.